File: Problems/187. Repeated DNA Sequences/Solution.java
import java.util.*;

class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        HashMap<String, Integer> sequences = new HashMap<>();
        List<String> repeated_sequences = new LinkedList<>();

        for (int i = 0; i <= s.length() - 10; ++i) {
            String dna_sequence = s.substring(i, i + 10);

            int num_occurences = sequences.getOrDefault(dna_sequence, 0);
            if (num_occurences == 1) {
                repeated_sequences.add(dna_sequence);
            }

            sequences.put(dna_sequence, num_occurences + 1);
        }

        return repeated_sequences;
    }
}

File: Problems/187. Repeated DNA Sequences/Solution.scala
import scala.collection.mutable

object Solution {
    val SEQUENCE_SIZE : Int = 10

    def findRepeatedDnaSequences(s: String): List[String] = {
        val sequence_counts = new mutable.HashMap[String, Int]()

        for (i <- 0 to (s.length - SEQUENCE_SIZE)) {
            val sequence : String = s.substring(i, i + SEQUENCE_SIZE)
            sequence_counts(sequence) = sequence_counts.getOrElse(sequence, 0) + 1
        }

        sequence_counts.filter(_._2 > 1).keys.toList
    }
}
